[summary] Time interpolation for 500 years crashes

Summary files going beyond year 2286 is supported, but if such a file is attempted interpolated in time, an error occurs:

```
        # Try with time interpolation involved:
>       dframe = summary.df(eclsum, time_index="yearly")

/home/berland/projects/ecl2df/tests/test_summary.py:306: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/home/berland/projects/ecl2df/ecl2df/summary.py:293: in df
    time_index_arg = resample_smry_dates(
/home/berland/projects/ecl2df/ecl2df/summary.py:183: in resample_smry_dates
    end_n = offset.rollforward(end_smry.date()).date()
pandas/_libs/tslibs/offsets.pyx:607: in pandas._libs.tslibs.offsets.BaseOffset.rollforward
    ???
pandas/_libs/tslibs/timestamps.pyx:1095: in pandas._libs.tslibs.timestamps.Timestamp.__new__
    ???
pandas/_libs/tslibs/conversion.pyx:430: in pandas._libs.tslibs.conversion.convert_to_tsobject
    ???
pandas/_libs/tslibs/conversion.pyx:507: in pandas._libs.tslibs.conversion.convert_datetime_to_tsobject
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   pandas._libs.tslibs.np_datetime.OutOfBoundsDatetime: Out of bounds nanosecond timestamp: 2499-12-31 00:00:00

pandas/_libs/tslibs/np_datetime.pyx:113: OutOfBoundsDatetime
```

--- a/ecl2df/summary.py
+++ b/ecl2df/summary.py
@@ -8,7 +8,7 @@ from typing import Dict, List, Optional, Union, Any
 # The name 'datetime' is in use by a function argument:
 import datetime as dt
 
-import dateutil.parser
+import dateutil
 import numpy as np
 import pandas as pd
 
@@ -52,7 +52,12 @@ def date_range(start_date: dt.date, end_date: dt.date, freq: str) -> List[dt.dat
     Returns:
         list of datetimes
     """
-    return pd.date_range(start_date, end_date, freq=PD_FREQ_MNEMONICS.get(freq, freq))
+    try:
+        return pd.date_range(
+            start_date, end_date, freq=PD_FREQ_MNEMONICS.get(freq, freq)
+        )
+    except pd.errors.OutOfBoundsDatetime:
+        return _fallback_date_range(start_date, end_date, freq)
 
 
 def _ensure_date_or_none(some_date: Optional[Union[str, dt.date]]) -> Optional[dt.date]:
@@ -74,7 +79,7 @@ def _ensure_date_or_none(some_date: Optional[Union[str, dt.date]]) -> Optional[d
     if some_date == "":
         return None
     if isinstance(some_date, str):
-        return dateutil.parser.parse(some_date).date()
+        return dateutil.parser.parse(some_date).date()  # type: ignore
     raise TypeError(f"Not a date type: {str(some_date)}")
 
 
@@ -118,6 +123,68 @@ def _crop_datelist(
     if isinstance(freq, (dt.date, dt.datetime)):
         return [freq]
     raise ValueError(f"Expected freq to an accepted string or datetime type was {freq}")
+
+
+def _fallback_date_roll(rollme: dt.datetime, direction: str, freq: str) -> dt.datetime:
+    """Fallback function for rolling dates forward or backward onto a
+    date frequency boundary.
+
+    This function reimplements pandas' DateOffset.roll_forward() and backward()
+    only for monthly and yearly frequency. This is necessary as Pandas does not
+    support datetimes beyond year 2262 due to all datetimes in Pandas being
+    represented by nanosecond accuracy.
+
+    This function is a fallback only, to keep support for using all Pandas timeoffsets
+    in situations where years beyond 2262 is not a issue."""
+    if direction not in ["back", "forward"]:
+        raise ValueError(f"Unknown direction {direction}")
+
+    if freq == "yearly":
+        if direction == "forward":
+            if rollme <= dt.datetime(year=rollme.year, month=1, day=1):
+                return dt.datetime(year=rollme.year, month=1, day=1)
+            return dt.datetime(year=rollme.year + 1, month=1, day=1)
+        else:
+            return dt.datetime(year=rollme.year, month=1, day=1)
+
+    if freq == "monthly":
+        if direction == "forward":
+            if rollme <= dt.datetime(year=rollme.year, month=rollme.month, day=1):
+                return dt.datetime(year=rollme.year, month=rollme.month, day=1)
+            return dt.datetime(
+                year=rollme.year, month=rollme.month, day=1
+            ) + dateutil.relativedelta.relativedelta(  # type: ignore
+                months=1
+            )
+        else:
+            return dt.datetime(year=rollme.year, month=rollme.month, day=1)
+
+    raise ValueError(
+        "Only yearly or monthly frequencies are "
+        "supported for simulations beyond year 2262"
+    )
+
+
+def _fallback_date_range(start: dt.date, end: dt.date, freq: str) -> List[dt.datetime]:
+    """Fallback routine for generating date ranges beyond Pandas datetime64[ns]
+    year-2262 limit.
+
+    Assumes that the start and end times already fall on a frequency boundary.
+    """
+    if freq == "yearly":
+        return [
+            dt.datetime(year=year, month=1, day=1)
+            for year in range(start.year, end.year + 1)
+        ]
+    if freq == "monthly":
+        dates = []
+        date = dt.datetime.combine(start, dt.datetime.min.time())
+        enddatetime = dt.datetime.combine(end, dt.datetime.min.time())
+        while date <= enddatetime:
+            dates.append(date)
+            date = date + dateutil.relativedelta.relativedelta(months=1)  # type: ignore
+        return dates
+    raise ValueError("Unsupported frequency for datetimes beyond year 2262")
 
 
 def resample_smry_dates(
@@ -164,7 +231,7 @@ def resample_smry_dates(
 
     # In case freq is an ISO-date(time)-string, interpret as such:
     try:
-        parseddate = dateutil.parser.isoparse(freq)
+        parseddate = dateutil.parser.isoparse(freq)  # type: ignore
         return [parseddate]
     except ValueError:
         # freq is a frequency string or datetime.date (or similar)
@@ -179,8 +246,16 @@ def resample_smry_dates(
     # will be mapped to [1997-11-01, 2020-04-01]
     # For yearly frequency it will return [1997-01-01, 2021-01-01].
     offset = pd.tseries.frequencies.to_offset(PD_FREQ_MNEMONICS.get(freq, freq))
-    start_n = offset.rollback(start_smry.date()).date()
-    end_n = offset.rollforward(end_smry.date()).date()
+    try:
+        start_n = offset.rollback(start_smry.date()).date()
+    except pd.errors.OutOfBoundsDatetime:
+        # Pandas only supports datetime up to year 2262
+        start_n = _fallback_date_roll(start_smry, "back", freq).date()
+    try:
+        end_n = offset.rollforward(end_smry.date()).date()
+    except pd.errors.OutOfBoundsDatetime:
+        # Pandas only supports datetime up to year 2262
+        end_n = _fallback_date_roll(end_smry, "forward", freq).date()
 
     if start_date is None:
         if normalize:
@@ -446,7 +521,10 @@ def _fix_dframe_for_libecl(dframe: pd.DataFrame) -> pd.DataFrame:
             # Do not use pd.Series.apply() here, Pandas would try to convert it to
             # datetime64[ns] which is limited at year 2262.
             dframe["DATE"] = pd.Series(
-                [dateutil.parser.parse(datestr) for datestr in dframe["DATE"]],
+                [
+                    dateutil.parser.parse(datestr)  # type: ignore
+                    for datestr in dframe["DATE"]
+                ],
                 dtype="object",
                 index=dframe.index,
             )

--- a/tests/test_summary.py
+++ b/tests/test_summary.py
@@ -1,6 +1,7 @@
 import os
 import sys
 import datetime
+from datetime import datetime as dt
 import logging
 from pathlib import Path
 
@@ -294,11 +295,11 @@ def test_foreseeable_future(tmpdir):
     assert (
         dframe.index
         == [
-            datetime.datetime(2000, 1, 1),
+            dt(2000, 1, 1),
             # This discrepancy is due to seconds as a 32-bit float
             # having an accuracy limit (roundoff-error)
             # https://github.com/equinor/ecl/issues/803
-            datetime.datetime(2499, 12, 31, 23, 55, 44),
+            dt(2499, 12, 31, 23, 55, 44),
         ]
     ).all()
 
@@ -312,7 +313,7 @@ def test_foreseeable_future(tmpdir):
     eclsum = df2eclsum(src_dframe, casename="PLUGABANDON")
     dframe = summary.df(eclsum)
     # Still buggy:
-    assert dframe.index[-1] == datetime.datetime(2068, 12, 31, 23, 57, 52)
+    assert dframe.index[-1] == dt(2068, 12, 31, 23, 57, 52)
 
     # Try with one-year timesteps, starting late:
     src_dframe = pd.DataFrame(
@@ -324,7 +325,7 @@ def test_foreseeable_future(tmpdir):
     eclsum = df2eclsum(src_dframe, casename="PLUGABANDON")
     dframe = summary.df(eclsum)
     # Works fine when stepping only 68 years:
-    assert dframe.index[-1] == datetime.datetime(2468, 1, 1, 0, 0, 0)
+    assert dframe.index[-1] == dt(2468, 1, 1, 0, 0, 0)
 
 
 def test_resample_smry_dates():
@@ -347,14 +348,12 @@ def test_resample_smry_dates():
     weekly = resample_smry_dates(ecldates, freq="weekly")
     assert len(weekly) == 159
 
-    assert resample_smry_dates(ecldates, freq="2001-04-05") == [
-        datetime.datetime(2001, 4, 5, 0, 0)
-    ]
+    assert resample_smry_dates(ecldates, freq="2001-04-05") == [dt(2001, 4, 5, 0, 0)]
     assert resample_smry_dates(ecldates, freq=datetime.date(2001, 4, 5)) == [
         datetime.date(2001, 4, 5)
     ]
-    assert resample_smry_dates(ecldates, freq=datetime.datetime(2001, 4, 5, 0, 0)) == [
-        datetime.datetime(2001, 4, 5, 0, 0)
+    assert resample_smry_dates(ecldates, freq=dt(2001, 4, 5, 0, 0)) == [
+        dt(2001, 4, 5, 0, 0)
     ]
     # start and end should be included:
     assert (
@@ -507,39 +506,33 @@ def test_smry_meta_synthetic():
         # # # # # # # # # # # # # # # # # # # # # # # #
         (
             pd.DataFrame([{"DATE": "2516-01-01", "FOPT": 1}]),
-            pd.DataFrame(
-                [{"FOPT": 1}], index=[datetime.datetime(2516, 1, 1)]
-            ).rename_axis("DATE"),
+            pd.DataFrame([{"FOPT": 1}], index=[dt(2516, 1, 1)]).rename_axis("DATE"),
         ),
         # # # # # # # # # # # # # # # # # # # # # # # #
         (
             pd.DataFrame([{"DATE": np.datetime64("2016-01-01"), "FOPT": 1}]),
-            pd.DataFrame(
-                [{"FOPT": 1}], index=[datetime.datetime(2016, 1, 1)]
-            ).rename_axis("DATE"),
+            pd.DataFrame([{"FOPT": 1}], index=[dt(2016, 1, 1)]).rename_axis("DATE"),
         ),
         # # # # # # # # # # # # # # # # # # # # # # # #
         (
             pd.DataFrame(
                 [{"DATE": pd.Timestamp(datetime.date(2016, 1, 1)), "FOPT": 1}]
             ),
-            pd.DataFrame(
-                [{"FOPT": 1}], index=[datetime.datetime(2016, 1, 1)]
-            ).rename_axis("DATE"),
+            pd.DataFrame([{"FOPT": 1}], index=[dt(2016, 1, 1)]).rename_axis("DATE"),
         ),
         # # # # # # # # # # # # # # # # # # # # # # # #
         (
             pd.DataFrame(
                 [
                     {
-                        "DATE": datetime.datetime(2016, 1, 1, 12, 34, 56),
+                        "DATE": dt(2016, 1, 1, 12, 34, 56),
                         "FOPT": 1,
                     }
                 ]
             ),
-            pd.DataFrame(
-                [{"FOPT": 1}], index=[datetime.datetime(2016, 1, 1, 12, 34, 56)]
-            ).rename_axis("DATE"),
+            pd.DataFrame([{"FOPT": 1}], index=[dt(2016, 1, 1, 12, 34, 56)]).rename_axis(
+                "DATE"
+            ),
         ),
         # # # # # # # # # # # # # # # # # # # # # # # #
         (
@@ -551,17 +544,13 @@ def test_smry_meta_synthetic():
         # # # # # # # # # # # # # # # # # # # # # # # #
         (
             pd.DataFrame([{"DATE": datetime.date(2017, 1, 1), "FOPT": 1}]),
-            pd.DataFrame(
-                [{"FOPT": 1}], index=[datetime.datetime(2017, 1, 1)]
-            ).rename_axis("DATE"),
+            pd.DataFrame([{"FOPT": 1}], index=[dt(2017, 1, 1)]).rename_axis("DATE"),
         ),
         # # # # # # # # # # # # # # # # # # # # # # # #
         (
             # Dates that go beyond Pandas' datetime64[ns] limits:
             pd.DataFrame([{"DATE": datetime.date(2517, 1, 1), "FOPT": 1}]),
-            pd.DataFrame(
-                [{"FOPT": 1}], index=[datetime.datetime(2517, 1, 1)]
-            ).rename_axis("DATE"),
+            pd.DataFrame([{"FOPT": 1}], index=[dt(2517, 1, 1)]).rename_axis("DATE"),
         ),
         # # # # # # # # # # # # # # # # # # # # # # # #
         (

--- a/tests/test_summary.py
+++ b/tests/test_summary.py
@@ -21,6 +21,8 @@ from ecl2df.summary import (
     df,
     smry_meta,
     _fix_dframe_for_libecl,
+    _fallback_date_roll,
+    date_range,
 )
 
 TESTDIR = Path(__file__).absolute().parent
@@ -83,7 +85,7 @@ def test_df_column_keys():
     assert set(sumdf.attrs["meta"].keys()) == {"FOPT", "FOPR"}
 
 
-def test_summary2df_dates(caplog):
+def test_summary2df_dates():
     """Test that we have some API possibilities with ISO dates"""
     eclfiles = EclFiles(DATAFILE)
 
@@ -303,6 +305,11 @@ def test_foreseeable_future(tmpdir):
         ]
     ).all()
 
+    # Try with time interpolation involved:
+    dframe = summary.df(eclsum, time_index="yearly")
+    assert len(dframe) == 501
+    assert dframe.index.max() == datetime.date(year=2500, month=1, day=1)
+
     # Try with one-year timesteps:
     src_dframe = pd.DataFrame(
         {
@@ -326,6 +333,155 @@ def test_foreseeable_future(tmpdir):
     dframe = summary.df(eclsum)
     # Works fine when stepping only 68 years:
     assert dframe.index[-1] == dt(2468, 1, 1, 0, 0, 0)
+
+
+@pytest.mark.parametrize(
+    "rollme, direction, freq, expected",
+    [
+        (
+            dt(3000, 1, 1),
+            "forward",
+            "yearly",
+            dt(3000, 1, 1),
+        ),
+        (
+            dt(3000, 1, 1),
+            "forward",
+            "monthly",
+            dt(3000, 1, 1),
+        ),
+        (
+            dt(3000, 1, 2),
+            "forward",
+            "yearly",
+            dt(3001, 1, 1),
+        ),
+        (
+            dt(3000, 1, 2),
+            "forward",
+            "monthly",
+            dt(3000, 2, 1),
+        ),
+        (
+            dt(3000, 1, 1),
+            "back",
+            "yearly",
+            dt(3000, 1, 1),
+        ),
+        (
+            dt(3000, 1, 1),
+            "back",
+            "monthly",
+            dt(3000, 1, 1),
+        ),
+        (
+            dt(3000, 12, 31),
+            "back",
+            "yearly",
+            dt(3000, 1, 1),
+        ),
+        (
+            dt(3000, 2, 2),
+            "back",
+            "monthly",
+            dt(3000, 2, 1),
+        ),
+        pytest.param(
+            dt(3000, 2, 2),
+            "forward",
+            "daily",
+            None,
+            marks=pytest.mark.xfail(raises=ValueError),
+        ),
+        pytest.param(
+            dt(3000, 2, 2),
+            "upwards",
+            "yearly",
+            None,
+            marks=pytest.mark.xfail(raises=ValueError),
+        ),
+    ],
+)
+def test_fallback_date_roll(rollme, direction, freq, expected):
+    assert _fallback_date_roll(rollme, direction, freq) == expected
+
+
+@pytest.mark.parametrize(
+    "start, end, freq, expected",
+    [
+        (
+            dt(3000, 1, 1),
+            dt(3002, 1, 1),
+            "yearly",
+            [
+                dt(3000, 1, 1),
+                dt(3001, 1, 1),
+                dt(3002, 1, 1),
+            ],
+        ),
+        (
+            dt(2999, 11, 1),
+            dt(3000, 2, 1),
+            "monthly",
+            [
+                dt(2999, 11, 1),
+                dt(2999, 12, 1),
+                dt(3000, 1, 1),
+                dt(3000, 2, 1),
+            ],
+        ),
+        pytest.param(
+            dt(3000, 1, 1),
+            dt(3000, 2, 1),
+            "weekly",
+            None,
+            marks=pytest.mark.xfail(raises=ValueError),
+        ),
+        (
+            # Crossing the problematic time boundary:
+            dt(2260, 1, 1),
+            dt(2263, 1, 1),
+            "yearly",
+            [
+                dt(2260, 1, 1),
+                dt(2261, 1, 1),
+                dt(2262, 1, 1),
+                dt(2263, 1, 1),
+            ],
+        ),
+        (
+            dt(3000, 1, 1),
+            dt(3000, 1, 1),
+            "yearly",
+            [
+                dt(3000, 1, 1),
+            ],
+        ),
+        (
+            dt(2000, 1, 1),
+            dt(2000, 1, 1),
+            "yearly",
+            [
+                dt(2000, 1, 1),
+            ],
+        ),
+        (
+            dt(2000, 1, 1),
+            dt(1000, 1, 1),
+            "yearly",
+            [],
+        ),
+        (
+            dt(3000, 1, 1),
+            dt(2000, 1, 1),
+            "yearly",
+            [],
+        ),
+    ],
+)
+def test_date_range(start, end, freq, expected):
+    # When dates are beyond year 2262, the function _fallback_date_range() is triggered.
+    assert date_range(start, end, freq) == expected
 
 
 def test_resample_smry_dates():

